# **Lab2: Indexing and Subsetting (20 points)**

#### **WRITE YOUR NAME HERE**


---
**Instructions**: Complete the lab below.
- <font color=red>Tasks</font> indicated in red must be completed to receive credit
- Include all code for your work
- Include markdown cells with written answers explaining your work when prompted
- Your completed lab must be submitted to Canvas as a pdf file by the end of the lab session
---




<br>




####<font color=red>Task 1</font>  (2 points)

The code below loads and preprocesses the US Census data described in lecture
* Run the code cell
* View the first few rows using `head()` to ensure the data has been properly reformatted for analysis


In [ ]:
# create url to census data
census_url <- url("https://www2.census.gov/programs-surveys/popest/tables/2010-2011/state/totals/nst-est2011-01.csv")

# load the census data into memory
census <- read.csv(census_url)

# remove unnecessary columns and rows
nb_row <- nrow(census)
census <- census[-c(1:8, (nb_row - 6):nb_row), -(6:10)]

# overwrite the column names using meaningful column names
names(census) <- c("state", "april2010census", "april2010base", "july2010pop", "july2011pop")

# remove periods from state names
census$state <- gsub('\\.', '', census$state)

# remove commas from populations and convert to numeric
census$april2010census <- as.numeric(gsub(',', '', census$april2010census))
census$april2010base <- as.numeric(gsub(',', '', census$april2010base))
census$july2010pop <- as.numeric(gsub(',', '', census$july2010pop))
census$july2011pop <- as.numeric(gsub(',', '', census$july2011pop))

# remove existing row names
rownames(census) <- NULL

In [ ]:
# View the first few rows
head(census)

# looking good!

state      april2010census april2010base july2010pop july2011pop
1 Alabama     4779736         4779735       4785401     4802740   
2 Alaska       710231          710231        714146      722718   
3 Arizona     6392017         6392013       6413158     6482505   
4 Arkansas    2915918         2915921       2921588     2937979   
5 California 37253956        37253956      37338198    37691912   
6 Colorado    5029196         5029196       5047692     5116796

<br>

---

<br>






####<font color=red>Task 2</font>  (4 points)

Use R code to answer the following. Also answer in complete sentences using text boxes where relevant.

1. What is the average population across states in July 2010?
2. What is the average population across states in July 2011?
3. Was there an increase or decrease in the average population?

In [ ]:
print("July 2010")
mean(census$july2010pop)

print("July 2011")
mean(census$july2011pop)


[1] "July 2010"


[1] 6065298

[1] "July 2011"


[1] 6109645

1. The average state population in July 2010 was approximately 6,065,298 people.
2. The average state population in July 2011 was approximately 6,109,645 people.
3. There was an average increase in state population from July 2010 to 2011.

<br>

---

<br>





####<font color=red>Task 3</font>  (2 points)
Use R code to answer the following. Also answer in complete sentences using text boxes where relevant.


1. What is the largest population in July 2010? [Hint: Use `max()`]

The function `which.max()` provides you with the index of the maximum. For example, the output of `which.max(c(99, 256, 48))` is 2 since the second entry of the vector, 256, is the maximum.
2. Use `which.max()` to find the state with the largest population.

In [ ]:
# 1.
max(census$july2010pop)

# 2. 
census[which.max(census$july2010pop), ] # find the row
# or
census$state[which.max(census$july2010pop)] # find the state

[1] 37338198

state      april2010census april2010base july2010pop july2011pop
5 California 37253956        37253956      37338198    37691912

[1] "California"

1. The largest state population in July 2010 is 37,253,956 people.
2. California is the state with the largest population in 2010.

<br>

---

<br>






####<font color=red>Task 4</font>  (2 points)

Use R code to answer the following. Also answer in complete sentences using text boxes where relevant.
1. Use `which.min()` to find the state with the smallest population
2. What is the population in this state?

In [ ]:
# 1.
census[which.min(census$july2010pop), ] # find the row
# or
census$state[which.min(census$july2010pop)] # find the state

# 2.
census[which.min(census$july2010pop), ]
# or
census[which.min(census$july2010pop), ]$july2010pop

state   april2010census april2010base july2010pop july2011pop
51 Wyoming 563626          563626        564554      568158

[1] "Wyoming"

state   april2010census april2010base july2010pop july2011pop
51 Wyoming 563626          563626        564554      568158

[1] 564554

1. Wyoming is the state with the smallest population.
2. The population in Wyoming in July 2010 is 564,554 people.

<br>

---

<br>






####<font color=red>Task 5</font>  (4 points)

The function `order()` returns the indexes of a sorted vector. For example, the output of `order(c(99, 256, 48), decreasing=FALSE)` is the vector `(3, 1, 2)` since the third number is the smallest (48), the first number is next (99), and the second number is the largest (256).

The example below shows how to sort a single vector:
```
x <- c(99, 256, 48)
x[order(x)]
```
<br>

Use R code to answer the following. Also answer in complete sentences using text boxes where relevant.

1. Sort the dataframe `census` by population in July 2010 from smallest to largest and store as another dataframe called `census_sorted`
2. Show the first few lines to ensure it is correct
3. Using the sorted dataframe, which areas of the US have the smallest populations? Why?
4. Using the sorted dataframe, which areas of the US have the largest populations? Why?

In [ ]:
# 1.
census_sorted <- census[order(census$july2010pop), ]

# 2. 
head(census_sorted)
tail(census_sorted)


state                april2010census april2010base july2010pop july2011pop
51 Wyoming              563626          563626        564554      568158     
9  District of Columbia 601723          601723        604912      617996     
46 Vermont              625741          625741        625909      626431     
35 North Dakota         672591          672591        674629      683932     
2  Alaska               710231          710231        714146      722718     
42 South Dakota         814180          814180        816598      824082

state        april2010census april2010base july2010pop july2011pop
39 Pennsylvania 12702379        12702379      12717722    12742886   
14 Illinois     12830632        12830632      12841980    12869257   
10 Florida      18801310        18801311      18838613    19057542   
33 New York     19378102        19378104      19395206    19465197   
44 Texas        25145561        25145561      25253466    25674681   
5  California   37253956        37253956      37338198    37691912

3. Aside from DC, states with the smallest populations tend to be landlocked and in the northern United States.
4. States with the larger populations tend to be larger in land area and/or bordering a coast (lake/ocean).

What other attributes do these states have?

<br>

---

<br>






####<font color=red>Task 6</font>  (2 points)

Use R code to answer the following. Also answer in complete sentences using text boxes where relevant.

1. Create a dataframe of states with populations between 1 and 2 million people in July 2010 and store as `census_1to2`
2. Among states with populations between 1 and 2 million people, which state has the largest population? Use R code to answer this question. [Hint: `which.max()`]


In [ ]:
# 1. 
census_1to2 <- census[ census$july2010pop > 1000000 & census$july2010pop < 2000000 , ]
census_1to2

# 2. 
census_1to2[which.max(census_1to2$july2010pop), ]


state         april2010census april2010base july2010pop july2011pop
12 Hawaii        1360301         1360301       1363359     1374810    
13 Idaho         1567582         1567582       1571102     1584985    
20 Maine         1328361         1328361       1327379     1328188    
28 Nebraska      1826341         1826341       1830141     1842641    
30 New Hampshire 1316470         1316472       1316807     1318194    
40 Rhode Island  1052567         1052567       1052528     1051302    
49 West Virginia 1852994         1852996       1854368     1855364

state         april2010census april2010base july2010pop july2011pop
49 West Virginia 1852994         1852996       1854368     1855364

Among states with populations between 1 and 2 million people, West Virginia has the largest population.

<br>

---

<br>






####<font color=red>Task 7</font>  (4 points)

Use R code to answer the following. Also answer in complete sentences using text boxes where relevant.

1. Calculate the difference in population for each state between 2010 and 2011. [Hint: population 2011 - population 2010]
2. Store as another variable called `pop_diff` in the `census` dataframe
3. How many states decreased in population between 2010 and 2011?
4. How many states had population increases less than 10,000 between 2010 and 2011?


In [ ]:
# 1. 
census$july2011pop - census$july2010pop

# 2. 
census$pop_diff <- census$july2011pop - census$july2010pop

# 3. 
nrow(census[census$pop_diff < 0, ])
# or 
length(census$pop_diff[census$pop_diff < 0])
# or
sum( census$pop_diff < 0 )

# 4. 
nrow(census[census$pop_diff > 0 & census$pop_diff < 10000, ])
# or 
length(census$pop_diff[ census$pop_diff > 0 & census$pop_diff < 10000 ])
# or
sum( census$pop_diff > 0 & census$pop_diff < 10000 )

[1]  17339   8572  69347  16391 353714  69104   5211   7343  13084 218929
[11] 103053  11451  13883  27277  26300  12107  12095  22133  29493    809
[21]  42608  32070   -956  34203   8440  14973   7241  12500  19039   1387
[31]  21562  16311  69991  96167   9303   6983  31324  33527  25164  -1226
[41]  42124   7484  45917 421215  41743    522  72651  87088    996  20108
[51]   3604

[1] 2

[1] 2

[1] 2

[1] 13

[1] 13

[1] 13

3. Two states decreased in population between July 2010 and July 2011
4. Thirteen states had population increases that were less than 10,000 from July 2010 t0 July 2011.